File: 2. Add Two Numbers/addtwonumbers.py
#!/usr/bin/env python3

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        current = dummy
        carry = 0
        while l1 or l2 or carry:
            sum = carry
            if l1:
                sum += l1.val
                l1 = l1.next
            if l2:
                sum += l2.val
                l2 = l2.next
            carry, val = divmod(sum, 10)
            current.next = ListNode(val)
            current = current.next
        return dummy.next

File: 2. Add Two Numbers/addtwonumbers.cpp
#include "addtwonumbers.h"
#include <__config>


ListNode* Solution::addTwoNumbers(ListNode* l1, ListNode* l2){

    ListNode* l = NULL;
    int carry = 0;
    while(l1 != NULL || l2 != NULL){
        int val1 = l1 == NULL ? 0 : l1->val;
        int val2 = l2 == NULL ? 0 : l2->val;
        int sum = val1 + val2 + carry;
        carry = sum / 10;
        l = addNode(l, sum % 10);
        if (l1 != NULL) l1 = l1->next;
        if (l2 != NULL) l2 = l2->next;
    }
    if (carry > 0){
        l = addNode(l, carry);
    }
    return l;
}


ListNode* Solution::addNode(ListNode* l, int val){
    if (l == NULL){
        l = new ListNode(val);
    }else{
        ListNode* p = l;
        while(p->next != NULL){
            p = p->next;
        }
        p->next = new ListNode(val);
    }
    return l;
}

void Solution::deleteNode(ListNode *l){

    ListNode* p = l;
    while(p != NULL){
        ListNode* q = p;
        p = p->next;
        delete q;
    }
}
